File: src/activity_heartbeat/activity_heartbeat_workflow.erl
-module(activity_heartbeat_workflow).

% elp:ignore W0012 W0040
-moduledoc false.

-export([
    execute/2
]).

-include_lib("temporal_sdk/include/workflow.hrl").

execute(_Context, [CancelationStage]) when is_integer(CancelationStage), CancelationStage > 0 ->
    A = start_activity(activity_heartbeat_activity, [], [{heartbeat_timeout, 1_000}]),
    case CancelationStage =< 3 of
        true ->
            %% Some other work is simulated by Temporal timer.
            start_timer(activity_heartbeat:stage_cancelation_time(CancelationStage), [wait]),
            #{state := canceled, details := D} = cancel_activity(A, [wait]),
            set_workflow_result(D);
        false ->
            wait(A),
            set_workflow_result(["activity closed"])
    end.

File: lib/activity_heartbeat/activity_heartbeat_workflow.ex
defmodule ActivityHeartbeat.Workflow do
  use TemporalSdk.Workflow
  @moduledoc false

  @impl true
  def execute(_context, [cancelation_stage])
      when is_integer(cancelation_stage) and cancelation_stage > 0 do
    a = start_activity(ActivityHeartbeat.Activity, [], heartbeat_timeout: 1_000)

    case cancelation_stage <= 3 do
      true ->
        # Some other work is simulated by Temporal timer.
        start_timer(ActivityHeartbeat.stage_cancelation_time(cancelation_stage), [:wait])
        %{state: :canceled, details: d} = cancel_activity(a, [:wait])
        set_workflow_result(d)

      false ->
        wait(a)
        set_workflow_result(["activity closed"])
    end
  end
end
